internal/hcp/internal/controllers/telemetrystate/controller.go

```diff
@@ -0,0 +1,168 @@
+}
+
+func (r *telemetryStateReconciler) Reconcile(ctx context.Context, rt controller.Runtime, req controller.Request) error {
+	// The runtime is passed by value so replacing it here for the remainder of this
+	// reconciliation request processing will not affect future invocations.
+	rt.Logger = rt.Logger.With("resource-id", req.ID, "controller", StatusKey)
+
+	rt.Logger.Trace("reconciling telemetry-state")
+
+	// First get the link resource in order to build a hcp client. If the link resource
+	// doesn't exist then the telemetry-state should not exist either.
+	res, err := getLinkResource(ctx, rt)
+	if err != nil {
+		rt.Logger.Error("failed to lookup Link resource", "error", err)
+		return err
+	}
+	if res == nil {
+		return r.ensureStateDeleted(ctx, rt)
+	}
+
+	// Check that the link resource indicates the cluster is linked
+	// If the cluster is not linked, the telemetry-state resource should not exist
+	if linked, reason := link.IsLinked(res.GetResource()); !linked {
+		rt.Logger.Trace("cluster is not linked", "reason", reason)
+		return r.ensureStateDeleted(ctx, rt)
+	}
+
+	hcpClient, err := r.hcpClientFn(res.Data)
+	if err != nil {
+		rt.Logger.Error("error creating HCP Client", "error", err)
+		return err
+	}
+
+	// Get the telemetry configuration and observability scoped credentials from hcp
+	tCfg, err := hcpClient.FetchTelemetryConfig(ctx)
+	if err != nil {
+		rt.Logger.Error("error requesting telemetry config", "error", err)
+		return err
+	}
+	clientID, clientSecret, err := hcpClient.GetObservabilitySecret(ctx)
+	if err != nil {
+		rt.Logger.Error("error requesting telemetry credentials", "error", err)
+		return nil
+	}
+
+	// TODO allow hcp client config override from hcp TelemetryConfig
+	hcpCfg := res.GetData().HcpConfig
+
+	// TODO implement proxy options from hcp
+	proxyCfg := &pbhcp.ProxyConfig{}
+
+	state := &pbhcp.TelemetryState{
+		ResourceId:   res.GetData().ResourceId,
+		ClientId:     clientID,
+		ClientSecret: clientSecret,
+		HcpConfig:    hcpCfg,
+		Proxy:        proxyCfg,
+		Metrics: &pbhcp.MetricsConfig{
+			Labels:   tCfg.MetricsConfig.Labels,
+			Disabled: tCfg.MetricsConfig.Disabled,
+		},
+	}
+
+	if tCfg.MetricsConfig.Endpoint != nil {
+		state.Metrics.Endpoint = tCfg.MetricsConfig.Endpoint.String()
+	}
+	if tCfg.MetricsConfig.Filters != nil {
+		state.Metrics.IncludeList = []string{tCfg.MetricsConfig.Filters.String()}
+	}
+
+	stateData, err := anypb.New(state)
+	if err != nil {
+		rt.Logger.Error("error marshalling telemetry-state data", "error", err)
+		return err
+	}
+
+	_, err = rt.Client.Write(ctx, &pbresource.WriteRequest{Resource: &pbresource.Resource{
+		Id: &pbresource.ID{
+			Name: "global",
+			Type: pbhcp.TelemetryStateType,
+		},
+		Data: stateData,
+	}})
+	if err != nil {
+		rt.Logger.Error("error updating telemetry-state", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+func (r *telemetryStateReconciler) ensureStateDeleted(ctx context.Context, rt controller.Runtime) error {
```

Comment:
does this need the `(r *telemetryStateReconciler)` receiver arg? seems like we could do without that unless there's a particular reason to have it.